lib/phoenix/socket/transport.ex

```diff
@@ -353,6 +353,75 @@ defmodule Phoenix.Socket.Transport do
     end
   end
 
+  @doc """
+  Checks the Websocket subprotocols request header against the allowed subprotocols.
+
+  Should be called by transports before connecting when appropriate.
+  If the sec-websocket-protocol header matches the allowed subprotocols,
+  it will put sec-websocket-protocol response header and return the given connection.
+  If no sec-websocket-protocol header was sent it will return the given connection.
+
+  Otherwise a 403 Forbidden response will be sent and the connection halted.
+  It is a noop if the connection has been halted.
+  """
+  def check_subprotocols(conn, subprotocols \\ nil)
+
+  def check_subprotocols(%Plug.Conn{halted: true} = conn, _subprotocols), do: conn
+  def check_subprotocols(conn, nil), do: conn
+
+  def check_subprotocols(conn, subprotocols) when is_list(subprotocols) do
+    import Plug.Conn
+    request_subprotocols =
+      conn
+      |> get_req_header("sec-websocket-protocol")
+      |> List.first
```

Comment:
The list can be empty, which means this would return nil. If get_req_header returns an empty list, then it mean we shoudn't do any subverification.